include/wil/network.h

```diff
@@ -0,0 +1,1372 @@
+        F function_table{};
+        void load() WI_NOEXCEPT;
+
+        // must guarantee Winsock does not unload while we have dynamically loaded function pointers
+        const ::wil::network::unique_wsacleanup_call wsa_reference_count;
+    };
+
+    //
+    // explicit specializations for socket_extension_function_table_t
+    //
+    template <>
+    inline socket_extension_function_table_t<WINSOCK_EXTENSION_FUNCTION_TABLE>::operator bool() const WI_NOEXCEPT
+    {
+        return function_table.AcceptEx != nullptr;
+    }
+
+    template <>
+    inline socket_extension_function_table_t<RIO_EXTENSION_FUNCTION_TABLE>::operator bool() const WI_NOEXCEPT
+    {
+        return function_table.RIOReceive != nullptr;
+    }
+
+    template <>
+    inline socket_extension_function_table_t<WINSOCK_SOCKET_NOTIFICATION_TABLE>::operator bool() const WI_NOEXCEPT
+    {
+        return function_table.ProcessSocketNotifications != nullptr;
+    }
+
+    template <>
+    inline void socket_extension_function_table_t<WINSOCK_EXTENSION_FUNCTION_TABLE>::load() WI_NOEXCEPT
+    {
+        // if WSAStartup failed, immediately exit
+        if (!wsa_reference_count)
+        {
+            return;
+        }
+
+        // we need a temporary socket for the IOCTL to load the functions
+        const ::wil::unique_socket localSocket{::socket(AF_INET, SOCK_DGRAM, IPPROTO_UDP)};
+        if (!localSocket)
+        {
+            return;
+        }
+
+        const auto load_function_pointer = [&](GUID extensionGuid, void* functionPtr) WI_NOEXCEPT {
+            DWORD unused_bytes;
+            const auto error{::WSAIoctl(
+                localSocket.get(),
+                SIO_GET_EXTENSION_FUNCTION_POINTER,
+                &extensionGuid,
+                DWORD{sizeof(extensionGuid)},
+                functionPtr,
+                DWORD{sizeof(functionPtr)},
+                &unused_bytes,
+                nullptr,
+                nullptr)};
+            return error == 0 ? S_OK : HRESULT_FROM_WIN32(::WSAGetLastError());
+        };
+
+        if (FAILED_LOG(load_function_pointer(WSAID_ACCEPTEX, &function_table.AcceptEx)) ||
+            FAILED_LOG(load_function_pointer(WSAID_CONNECTEX, &function_table.ConnectEx)) ||
+            FAILED_LOG(load_function_pointer(WSAID_DISCONNECTEX, &function_table.DisconnectEx)) ||
+            FAILED_LOG(load_function_pointer(WSAID_GETACCEPTEXSOCKADDRS, &function_table.GetAcceptExSockaddrs)) ||
+            FAILED_LOG(load_function_pointer(WSAID_TRANSMITFILE, &function_table.TransmitFile)) ||
+            FAILED_LOG(load_function_pointer(WSAID_TRANSMITPACKETS, &function_table.TransmitPackets)) ||
+            FAILED_LOG(load_function_pointer(WSAID_WSARECVMSG, &function_table.WSARecvMsg)) ||
+            FAILED_LOG(load_function_pointer(WSAID_WSASENDMSG, &function_table.WSASendMsg)))
+        {
+            // if any failed to be found, something is very broken
+            // all should load, or all should fail

```

Comment:
This comment is a little confusing for two reasons. The first is that "if any failed to be found, something is very broken" conflicts with "or all should fail" - if the second comment is true, the first comment should be more like "if any failed to be found _after the first function_". The second thing is that if what I said previously is what you intended to say, we probably don't want to `FAILED_LOG` if the first function pointer fails to load if that's an expected outcome.

Alternatively, if your first comment is the correct one, the second comment should probably just be deleted (and the `FAILED_LOG` is okay)

Reply:
Good point. I've updated the comment to be more precise. These really should never fail at all - but because Winsock has some ancient 3rd party providers that could be installed, they might end up failing to find these extension functions if they have some super-old 3rd party software that doesn't support this.

I think "FAILED_LOG" is correct for this usage.

Happy to hear your thoughts.

Reply:
Based off the updated comment, I'd agree - the log is fine then